doc/manual/src/language/scope.md

```diff
@@ -1,14 +1,34 @@
 # Scoping rules
 
-Nix is [statically scoped](https://en.wikipedia.org/wiki/Scope_(computer_science)#Lexical_scope), but with multiple scopes and shadowing rules.
+A _scope_ in the Nix language is an [identifier]-keyed dictionary, mapping each identifier to an expression and a _definition type_.
+The definition type is either _explicit_ or _implicit_.
```

Comment:
I don't understand what you're proposing here. If you meant to mark this line, what alternative would you propose for listing definition types, and where is ‘in place’?

Reply:
I would just list the definition types directly without classifying them into explicit and implicit, and explain the semantics of each in their own respective section/page.

Reply:
Hm. Perhaps I've written this confusingly.

The thing this paragraph is trying to express is that we have (conceptually):
```cpp
enum DefinitionType { Implicit, Explicit };
using Scope = std::map<Name, std::pair<Expr, DefinitionType>>;
```

It happens to be the case that the expressions that extend scopes can be classified into implicit and explicit, but that doesn't matter for purposes of defining what data a scope contains. A hypothetical Nix language extension could allow let-expressions to create both implicit and explicit definitions, for example:
```nix
let
  a = 1;
  implicit b = 2;
in ...
```

But there is a meaningful difference between the scopes $\\{ \texttt{\"a\"} \mapsto (1, \mathit{implicit}) \\}$ and $\\{ \texttt{"a"} \mapsto (1, \mathit{explicit}) \\}$; the difference is revealed by the fact that `with { a = 2; }; a` evaluates to 2 in the former scope and 1 in the latter. So the difference should be explained in the description of what a scope is.

The previous version of this documentation achieved this goal by partitioning scopes into primary and secondary maps, like so:
```cpp
struct Scope {
  std::map<Name, Expr> primary;
  std::map<Name, Expr> secondary;
}
```

This is isomorphic to the other `Scope` type, but when I tried explaining the concepts ‘primary’ and ‘secondary’, I noticed that they don't really connect to the language we naturally use to describe the special behavior of definitions created by `with`. We naturally describe those definitions as implicit, so I thought it would be easier to understand if I took the implicit/explicit distinction and applied it directly to definitions (map entries) in scopes.

Reply:
I understood what you were trying to achieve, but I think it front-loads a lot of complexity. IMO we have to design the manual for incremental reading, and putting too much operational detail into an overview will just destroy reader's motivation to plow through the rest. Therefore I'm suggesting working out the details in the respective definition types, and e.g. be appropriately more diligent when describing the semantics of `with`.

Reply:
I was thinking of this as a specification of what ‘scope’ means, not an overview.

There should be, somewhere, a reference-level specification of what a scope is, so that other places in the language reference can link to that when scopes are mentioned.

If not here, where should that specification go?

Reply:
I gave it another thought, and yes, that's probably the best way to do it. Thanks for the productive discussion and sorry for the delay.